<?xml version="1.0"?>
<launch>
    <!-- Set environment variables for software rendering and Qt -->
    <env name="LIBGL_ALWAYS_SOFTWARE" value="1"/>
    <env name="QT_QPA_PLATFORM" value="xcb"/>
    
    <!-- Set GAZEBO_MODEL_PATH -->
    <env name="GAZEBO_MODEL_PATH" value="$(find gazebo_world)/models"/>

    <!-- Launch arguments -->
    <arg name="paused"          default="true"/>   <!-- Start PAUSED to allow proper loading -->
    <arg name="use_sim_time"    default="true"/>
    <arg name="gui"             default="true"/>
    <arg name="headless"        default="false"/>
    <arg name="debug"           default="false"/>
    <arg name="verbose"         default="false"/>
    
    <!-- Rectangle walking parameters -->
    <arg name="side_long"       default="0.30"/>  <!-- 30cm -->
    <arg name="side_short"      default="0.20"/>  <!-- 20cm -->
    <arg name="num_laps"        default="2"/>     <!-- Number of rectangles to walk -->
    
    <!-- Robot spawn position (centered, facing forward) -->
    <arg name="x"               default="0"/>
    <arg name="y"               default="0"/>
    <arg name="z"               default="0.25"/>
    <arg name="roll"            default="0"/>
    <arg name="pitch"           default="0"/>
    <arg name="yaw"             default="0"/>  <!-- Facing forward (0 degrees) -->

    <!-- Set use_sim_time parameter -->
    <param name="/use_sim_time" value="$(arg use_sim_time)"/>

    <!-- Launch Gazebo with empty world (no line needed) -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name"      value="$(find gazebo_world)/worlds/line_follower.world"/>
        <arg name="paused"          value="$(arg paused)"/>
        <arg name="use_sim_time"    value="$(arg use_sim_time)"/>
        <arg name="gui"             value="$(arg gui)"/>
        <arg name="headless"        value="$(arg headless)"/>
        <arg name="debug"           value="$(arg debug)"/>
        <arg name="verbose"         value="$(arg verbose)"/>
    </include>

    <!-- Load robot description -->
    <param name="robot_description" command="$(find xacro)/xacro '$(find ainex_description)/urdf/ainex.xacro'" />

    <!-- Spawn robot model in Gazebo -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" 
          args="-urdf -param robot_description -model ainex
                -x $(arg x) -y $(arg y) -z $(arg z) 
                -R $(arg roll) -P $(arg pitch) -Y $(arg yaw)
                -J l_sho_roll -1.403 -J l_el_yaw -1.226 
                -J r_sho_roll 1.403 -J r_el_yaw 1.226" 
          respawn="false" output="screen" />

    <!-- Load position controller -->
    <include file="$(find ainex_gazebo)/launch/position_controller.launch"/>
    
    <!-- Load ainex kinematics controller for robot control -->
    <include file="$(find ainex_kinematics)/launch/ainex_controller.launch">
        <arg name="gazebo_sim" value="true"/>
    </include>

    <!-- Delayed start script - unpauses Gazebo after robot is loaded -->
    <node name="delayed_start" pkg="line_follower_test" type="delayed_start.sh" output="screen" />

    <!-- Rectangle Walker Node - Makes robot walk in rectangular pattern -->
    <node name="rectangle_walker" pkg="line_follower_test" type="rectangle_walker_node.py" output="screen"
          launch-prefix="bash -c 'sleep 5; $0 $@' ">
        <param name="side_long" value="$(arg side_long)"/>
        <param name="side_short" value="$(arg side_short)"/>
        <param name="num_laps" value="$(arg num_laps)"/>
        <param name="forward_speed" value="0.05"/>  <!-- 5cm per step - matches Lesson 4 tutorial (x_move_amplitude) -->
        <param name="turn_speed" value="10"/>  <!-- 10 degrees per step -->
    </node>

</launch>
